src/hypothesis/internal/conjecture/engine.py

```diff
@@ -1541,6 +1543,60 @@ def reoffset(o):
 
         minimize_int(offset, reoffset)
 
+    def shrink_offset_pairs(self):
+        """Lower any two blocks offset from each other the same ammount.
+
+        Try lowering every non-zero pair of non-shrinking blocks by a
+        common offset.
+
+        """
+        self.debug('Shrinking offset pairs.')
+
+        current = self.shrink_target
+        blocked = [current.buffer[u:v] for u, v in current.blocks]
+
+        n = len(blocked)
+
+        shrink_blocks = [b for b in range(n) if self.is_shrinking_block(b)]
+
+        ints = [int_from_bytes(blocked[i]) for i in range(n)]
+
+        nonzero = sorted([b for b in range(n)
+                          if ints[b] > 0 and b not in shrink_blocks])
+
+        if len(nonzero) == 0:
+            return
+
+        # TODO: Should we try combinations of > 2?
```

Comment:
We probably shouldn't!

The reason for the insistence on pairs is twofold:

1. Most non-trivial interactions seem to be between pairs of things, so it's very much the low-hanging fruit.
2. It gets very expensive very fast! Checking all pairs of things is just about feasible, but checking all k-combinations is `O(n^k)`, which gets slow fast.

Reply:
Good point. My thought here was that once we get the "expensive" passes, we have a much smaller number of non-zero blocks, so`k` is pretty low. But as you point out `n^k` gets pretty large beyond `k=2`.